courses/tasks.py

```diff
@@ -0,0 +1,66 @@
+"""
+Tasks for the courses app
+"""
+import logging
+from mitxpro.celery import app
+from mitxpro.utils import now_in_utc
+from courses.models import CourseRun, CourseRunCertificate
+from courses.utils import ensure_course_run_grade, process_course_run_grade_certificate
+from courseware.api import get_edx_grades_with_users
+
+log = logging.getLogger(__name__)
+
+
+@app.task
+def generate_course_certificates():
+    """
+    Task to generate certificates for courses.
+    """
+    now = now_in_utc()
+    course_runs = CourseRun.objects.filter(end_date__lt=now).exclude(
+        id__in=CourseRunCertificate.objects.values_list("course_run__id", flat=True)
+    )
+
+    for run in course_runs:
+        try:
+            edx_grade_user_iter = get_edx_grades_with_users(run)
```

Comment:
Sorry I didn't notice this before, but this exception handling won't work. `get_edx_grades_with_users` is a generator, so the exceptions are only raised when you're actually iterating. If we define a new generator, we can ignore any edX errors 

```python
from requests.exceptions import HTTPError
def exception_logging_generator(generator):
    """Returns a new generator that logs exceptions from the given generator and continues with iteration"""
    while True:
        try:
            yield next(generator)
        except StopIteration:
            return
        except Exception as e:
            log.exception("Error fetching user grades from edX")
# ...

@app.task
def generate_course_certificates():
  # ...
  edx_grade_user_gen = exception_logging_generator(get_edx_grades_with_users(run))
  # ...
  for edx_grade, user in edx_grade_user_gen:
    # ...
```

I think it's appropriate to just ignore

Reply:
I did not notice that before too and it makes sense here. 
but I am thinking why we are using `yield` in `get_edx_grades_with_users`. did it concern to memory before ?

Reply:
Just a python best practice. The fact that we can change the exception handling by defining this new generator is a nice bit of flexibility that we wouldn't get if `get_edx_grades_with_users` just returned a list or something